File: tools/AtlasBuildTool/BuildTarget/BuildTargetBase.cs
// Copyright(c) 2023-present, Atlas.
// Distributed under the MIT License (http://opensource.org/licenses/MIT)

namespace AtlasBuildTool.BuildTarget;

class BuildTargetException : Exception
{
    public BuildTargetException(string message) : base(message)
    {
    }
}

public abstract class BuildTargetBase
{
    public string TargetName { get; }
    
    public string RootDirectory { get; }

    public List<BuildTargetBase> PublicLinkBuildTargets { get; set; } = new List<BuildTargetBase>();
    
    public List<BuildTargetBase> PrivateLinkBuildTargets { get; set; } = new List<BuildTargetBase>();
    
    public HashSet<string> Options { get; } = new HashSet<string>();

    public Dictionary<string, string> OneValueArgs { get; } = new Dictionary<string, string>();

    public Dictionary<string, List<string>> MultiValueArgs { get; } = new Dictionary<string, List<string>>();

    protected BuildTargetBase(string targetName, string rootDirectory)
    {
        TargetName = targetName;
        RootDirectory = rootDirectory;
    }
    
    protected BuildTargetBase(string targetName, string rootDirectory, HashSet<string> options, Dictionary<string, string> oneValueArgs, Dictionary<string, List<string>> multiValueArgs)
    {
        TargetName = targetName;
        RootDirectory = rootDirectory;
        Options = options;
        OneValueArgs = oneValueArgs;
        MultiValueArgs = multiValueArgs;
    }

    public abstract bool NeedExportSymbol();
}

File: tools/atlas_builder/src/build_target/BuildTargetBase.kt
/*
 * Copyright(c) 2023-present, Atlas.
 * Distributed under the MIT License (http://opensource.org/licenses/MIT)
 */

package com.spacelion.atlas_builder.build_target

import java.io.File

abstract class BuildTargetBase(val name : String,
                      val rootDirectory : File,
                      val options : Set<String>,
                      val oneValueArgs : Map<String, String>,
                      val multiValueArgs : Map<String, List<String>>) {

    private val publicLinkTargets : MutableSet<BuildTargetBase> = mutableSetOf()
    private val privateLinkTargets : MutableSet<BuildTargetBase> = mutableSetOf()

    abstract fun exportSymbol() : Boolean

    fun getPublicLinkTargets() : Set<BuildTargetBase> = publicLinkTargets
    fun getPrivateLinkTargets() : Set<BuildTargetBase> = privateLinkTargets
    fun addPublicLinkTarget(target: BuildTargetBase)  = publicLinkTargets.add(target)
    fun addPrivateLinkTarget(target: BuildTargetBase) = privateLinkTargets.add(target)

    open fun getDependenciesBuildTargets(dependencies : MutableSet<BuildTargetBase>) {
        for (target in publicLinkTargets) {
            target.getDependenciesBuildTargets(dependencies)
            dependencies.add(target)
        }

        for (target in privateLinkTargets) {
            target.getDependenciesBuildTargets(dependencies)
            dependencies.add(target)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        check(other is BuildTargetBase)
        return name == other.name
    }

    override fun hashCode(): Int {
        return name.hashCode()
    }
}